File: backend/app/local/gpt_oauth.py
"""OAuth da assinatura ChatGPT / Codex CLI. Token em ~/.codex/auth.json."""

from __future__ import annotations

import json
import os
from pathlib import Path
from typing import Any

from app.local.cursor_state import jwt_expired, jwt_exp_unix


def auth_path(cfg: dict | None = None) -> Path:
    env_file = os.environ.get("CODEX_AUTH_PATH", "").strip()
    if env_file:
        return Path(env_file).expanduser()
    env_home = os.environ.get("CODEX_HOME", "").strip()
    if env_home:
        return Path(env_home).expanduser() / "auth.json"
    stored = ""
    if cfg:
        stored = str((cfg.get("paths") or {}).get("codex_auth") or "").strip()
    if stored:
        return Path(stored).expanduser()
    return Path.home() / ".codex" / "auth.json"


def parse_auth_blob(data: Any) -> tuple[str | None, str | None]:
    """Devolve (access_token, account_id). Não lê OPENAI_API_KEY."""
    if isinstance(data, str):
        text = data.strip()
        if not text:
            return None, None
        try:
            data = json.loads(text)
        except json.JSONDecodeError:
            return text, None
    if not isinstance(data, dict):
        return None, None
    tokens = data.get("tokens") if isinstance(data.get("tokens"), dict) else data
    token = str(tokens.get("access_token") or data.get("access_token") or "").strip()
    account_id = str(tokens.get("account_id") or data.get("account_id") or "").strip() or None
    if not token:
        return None, None
    return token, account_id


def from_auth_file(path: Path) -> tuple[str | None, str | None, str | None]:
    if not path.is_file():
        return None, None, None
    try:
        data = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError) as exc:
        return None, None, f"auth.json: {exc}"
    token, account_id = parse_auth_blob(data)
    if not token:
        return None, None, "auth.json sem access_token — rode `codex login`"
    return token, account_id, None


def gpt_token_candidates(cfg: dict | None = None) -> list[tuple[str, str, str | None, int | None]]:
    """Arquivo do Codex CLI. Token colado no painel não entra aqui."""
    found: list[tuple[str, str, str | None, int | None]] = []
    seen: set[str] = set()

    def add(source: str, token: str | None, account_id: str | None) -> None:
        if not token or token in seen:
            return
        seen.add(token)
        found.append((source, token, account_id, jwt_exp_unix(token)))

    token, account_id, _err = from_auth_file(auth_path(cfg))
    add("auth", token, account_id)
    return found


def gpt_missing_hint(cfg: dict | None = None) -> str:
    path = auth_path(cfg)
    if path.is_file():
        return "auth.json sem access_token — rode `codex login` neste Mac"
    return f"Nenhum login Codex encontrado — rode `codex login` neste Mac (sem {path})"


def gpt_token_expired(token: str) -> bool:
    return jwt_expired(token)

File: backend/src/local/gptOauth.ts
/**
 * OAuth da assinatura ChatGPT / Codex CLI. Token em ~/.codex/auth.json
 * Port of backend-python-legacy/app/local/gpt_oauth.py
 */
import { existsSync, readFileSync } from "node:fs";
import { homedir } from "node:os";
import { join } from "node:path";
import { jwtExpUnix, jwtExpired } from "./cursorState.js";

function expandUser(p: string): string {
  if (p === "~") return homedir();
  if (p.startsWith("~/")) return join(homedir(), p.slice(2));
  return p;
}

export function authPath(cfg?: Record<string, unknown> | null): string {
  const envFile = (process.env.CODEX_AUTH_PATH ?? "").trim();
  if (envFile) return expandUser(envFile);
  const envHome = (process.env.CODEX_HOME ?? "").trim();
  if (envHome) return join(expandUser(envHome), "auth.json");
  let stored = "";
  if (cfg) {
    const paths = cfg.paths as Record<string, unknown> | undefined;
    stored = String(paths?.codex_auth ?? "").trim();
  }
  if (stored) return expandUser(stored);
  return join(homedir(), ".codex", "auth.json");
}

export function parseAuthBlob(data: unknown): [string | null, string | null] {
  if (typeof data === "string") {
    const text = data.trim();
    if (!text) return [null, null];
    try {
      data = JSON.parse(text);
    } catch {
      return [text, null];
    }
  }
  if (data === null || typeof data !== "object" || Array.isArray(data)) return [null, null];
  const dict = data as Record<string, unknown>;
  const tokensRaw = dict.tokens;
  const tokens = tokensRaw !== null && typeof tokensRaw === "object" && !Array.isArray(tokensRaw)
    ? (tokensRaw as Record<string, unknown>)
    : (dict as Record<string, unknown>);
  const token = String(tokens.access_token ?? dict.access_token ?? "").trim();
  const accountIdRaw = tokens.account_id ?? dict.account_id;
  const accountId = String(accountIdRaw ?? "").trim() || null;
  if (!token) return [null, null];
  return [token, accountId];
}

export function fromAuthFile(path: string): [string | null, string | null, string | null] {
  if (!existsSync(path)) return [null, null, null];
  let data: unknown;
  try {
    data = JSON.parse(readFileSync(path, "utf-8"));
  } catch (e) {
    return [null, null, `auth.json: ${e}`];
  }
  const [token, accountId] = parseAuthBlob(data);
  if (!token) return [null, null, "auth.json sem access_token — rode `codex login`"];
  return [token, accountId, null];
}

export function gptTokenCandidates(
  cfg?: Record<string, unknown> | null,
): Array<[string, string, string | null, number | null]> {
  const found: Array<[string, string, string | null, number | null]> = [];
  const seen = new Set<string>();

  function add(source: string, token: string | null, accountId: string | null): void {
    if (!token || seen.has(token)) return;
    seen.add(token);
    found.push([source, token, accountId, jwtExpUnix(token)]);
  }

  const [token, accountId] = fromAuthFile(authPath(cfg));
  add("auth", token, accountId);
  return found;
}

export function gptMissingHint(cfg?: Record<string, unknown> | null): string {
  const p = authPath(cfg);
  if (existsSync(p)) return "auth.json sem access_token — rode `codex login` neste computador";
  return `Nenhum login Codex encontrado — rode \`codex login\` neste computador (sem ${p})`;
}

export function gptTokenExpired(token: string): boolean {
  return jwtExpired(token);
}
